### PR TITLE
Op/balazev

### PR DESCRIPTION
**Pull request by fedelopez77**
* Adapting models to Knowledge graph models, so they can represent more complex relations.
    * Adapting preprocessing so indexes of items and users are part of the same sequence
 * Adding Multi relational Poincare and Euclidean models (Balazevic et al https://arxiv.org/abs/1905.09791)
 * Adapting HingeLoss to be tested with these models
 * Adding Binary Cross Entropy Loss

Disclaimer: it says 49 commits because I branched off a branch that wasn't merged into master yet, but the real changes are way less.

### CODE REVIEW

File: hyperbolic-rs/preprocess.py [hunk 1]
```diff
@@ -43,23 +43,30 @@
...
 
-def map_item_ids_to_sequential_ids(samples):
+def map_raw_ids_to_sequential_ids(samples):
     """
     For each unique user or item id, this function creates a mapping to a sequence of number starting in 0.
     This will be the index of the embeddings in the model.
 
+    Items ids will be from 0 to n_items - 1.
+    Users ids will be from n_items to n_items + n_users - 1
+    This condition is required to later build the distance matrix
```
Comment (mscholz) [inline hyperbolic-rs/preprocess.py:53]: You probably want to generalize this to more types, and a varying number of types?

File: hyperbolic-rs/rudders/models/euclidean.py [hunk 2]
```diff
@@ -20,46 +20,61 @@
...
-            cosine_sim = tf.reduce_sum(user_embeds * item_embeds, axis=-1, keepdims=True)
-        return cosine_sim
+    def __init__(self, n_users, n_items, n_relations, item_ids, args):
+        super().__init__(n_users, n_items, n_relations, item_ids, args, train_bias=False)
+        self.relations = None
 
-    def distance(self, embed_a, embed_b, all_pairs):
-        """Distance based on cosine distance"""
-        return 1 - self.score(embed_a, embed_b, all_pairs)
+    def similarity_score(self, lhs, rhs, all_items):
```
Comment (mscholz) [inline hyperbolic-rs/rudders/models/euclidean.py:37]: What does the all_items do here?
Is this for item-to-item?

File: hyperbolic-rs/rudders/losses.py [hunk 3]
```diff
@@ -14,133 +14,271 @@
...
         pass
 
 
-class PairwiseHingeLoss(LossFunction):
-    """Pairwise ranking hinge loss."""
-
-    def __init__(self, n_users, n_items, args, **kwargs):
-        super(PairwiseHingeLoss, self).__init__(n_users, n_items, args)
-        self.gamma = tf.Variable(tf.keras.backend.ones(1), trainable=False)
+class TripletWithNegativeSampleLoss(LossFunction, abc.ABC):
```
Comment (mscholz) [inline hyperbolic-rs/rudders/losses.py:35]: Cool, I think I get the organization and logic of this.
But as discussed, since this is obsolete at this point I'd rather see the new version.

File: hyperbolic-rs/preprocess.py (hunk 1 above)
Comment (fedelopez77) [reply hyperbolic-rs/preprocess.py:53]: That is done in the new PR
